utility/service/service.go

```diff
@@ -122,19 +232,84 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
 	return nil
 }
 
-// TODO: implement
 // validateApplication makes sure the application has not received more relays than allocated in the current session.
-func (s servicer) validateApplication(meta *coreTypes.RelayMeta, session *coreTypes.Session) error {
-	/*
-		// if maxRelaysPerSession, overServiced := calculateAppSessionTokens(); overServiced {
-			return fmt.Errorf("application %s has exceeded its allocated relays %d for the session %d", meta.ApplicationPublicKey, maxRelaysPerSession)
-		}
-	*/
-	return nil
+func (s *servicer) validateApplication(session *coreTypes.Session, currentHeight int64) error {
+	// IMPROVE: use a function to get current height from the current session
+	servicerAppSessionTokens, err := s.calculateServicerAppSessionTokens(session, currentHeight)
+	if err != nil {
+		return fmt.Errorf("Error calculating servicer tokens for application: %w", err)
+	}
+
+	readCtx, err := s.GetBus().GetPersistenceModule().NewReadContext(currentHeight)
+	if err != nil {
+		return fmt.Errorf("Error getting read context: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	usedAppSessionTokens, err := readCtx.GetServicerTokenUsage(session)
+	if err != nil {
+		return fmt.Errorf("Error getting servicer token usage: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	if usedAppSessionTokens == nil || usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0 {
```

Comment:
I think we should log a warning for `usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0`

Reply:
Could you elaborate on this please? This condition is expected to occur whenever the application has sufficient tokens remaining for the session. Would a log message be used to track the relays?